### ISSUE TITLE
IWYU has issues on case-insensitive environments

### ISSUE THREAD

**Issue by dpunset** [include-what-you-use/include-what-you-use#213]
Hello!

I remember we talked about this quite some time ago but I found another issue when using IWYU and not respecting case-sensitiveness in the includes. I know you said that you preferred to keep IWYU case sensitive but it's quite an annoying problem using the tool in Windows when the codebase is already quite big. I wanted to expose the problem and an easy to solution that could help switch easily between case sensitive or case-insensitive.

So the problem I found now is that one of my includes was ending up two times in the lines list when reaching CalculateDesiredIncludesAndForwardDeclares. I saw that one character was upper-case and in the other it was lower-case, so I figured it was a case sensitiviness issue. I managed to fix that by changing a bit the code in OneIncludeOrForwardDeclareLine::matches(const string& quoted_include) but I still noticed that the include was not marked as desired, so I found other places where case was a problem. I proceed to make modifications on all the places I found and now it works. The places are:

OneIncludeOrForwardDeclareLine::matches
IwyuFileInfo::AssociatedDesiredIncludes()
CalculateDesiredIncludesAndForwardDeclares::include_map (every place this map is used)
CalculateDesiredIncludesAndForwardDeclares::associated_desired_includes (every place this set is used)

Basically in all these places I have to transform everything to lower_case before making a comparison.

I would like to suggest an easy way of switching easily between case-sensitive and insensitive. To achieve this I would replace all the places where you reference a header that you compare by a stringID (basically a crc, or uniqueID for the string). At the class that wraps this ID you could have two ways of computing this crc, a case-sensitive one and a non case-sensitive one. You can switch between the two based on environment flags, or command line or whatever you prefer without changing much the code.

What do you think?

**Comment by kimgr** [comment]
I remember last time we talked about this there were concerns about how to infer if the filesystem was indeed case-sensitive or not. Windows (and Mac, I believe) is case-insensitive by default, but can be changed to case-sensitive. Unix systems overall are typically case sensitive, but you can mount case-insensitive filesystems.

That said, I think it would be entirely appropriate to make the following assumptions:
- Windows - case-insensitive
- Mac - case-sensitive (I'm not sure about this, but I'm guessing Mac users are more likely to be writing portable code, respecting case)
- Other Unix systems - case-sensitive

and then add a command-line flag to override this default, e.g. --case-sensitive-filenames=true|false.

@vsapsai, what do you think? I remember you had concerns last time, but this would really lower the barrier to entry for Windows devs.

**Comment by vsapsai** [comment]
Let me check if my idea of fixing file paths covers this case. I am afraid the right fix is too involved. But according to places we need to touch

> OneIncludeOrForwardDeclareLine::matches
> IwyuFileInfo::AssociatedDesiredIncludes()
> CalculateDesiredIncludesAndForwardDeclares::include_map
> CalculateDesiredIncludesAndForwardDeclares::associated_desired_includes

my fears are exaggerated. Give me a week to check my idea.

For the reference, previous issue is #130.

**Comment by vsapsai** [comment]
@dpunset, I haven't finished checking where do we need to change string manipulation to file manipulation. Bad news is that I can already say we have a lot of code working with file paths and quoted includes as strings.

You can try to use `IncludeMapFile` like you've demonstrated in issue #215. But you'd better use only `IncludeMapFile`, not both `IncludeMapFile` and `IncludeMap`. Unfortunately I don't know how far it will get you.

**Comment by dpunset** [comment]
@vsapsai yeah I noticed. However I am confident a solution is totally feasible, right? In the end we only need a string for the case of the suggested header, but that's just for output and not to make any computation. I wanted to try to implement it myself but then got overloaded in my current project :(

**Comment by Hi-Angel** [comment]
Detecting case-sensitivity is wrong. You gonna end up in situation where a cross-platform app was checked on case-insensitive filesystem, and everything was fine; and then once changes commited upstream, it breaks for most other existing filesystems.

IWYU simply should not lower-case the include, ever.

**Comment by kimgr** [comment]
@Hi-Angel I can see that side of the argument, but look at it from the other side: if your codebase is Windows-only, you don't want to make lots of immaterial changes to satisfy IWYU.

For the cross-platform case, I think the answer is to prefer running IWYU on a case-insensitive platform.

But we might also want to consider letting users choose case sensitive or not for Windows, and just default to case-sensitive.

**Comment by Hi-Angel** [comment]
> @Hi-Angel I can see that side of the argument, but look at it from the other side: if your codebase is Windows-only, you don't want to make lots of immaterial changes to satisfy IWYU.

Fair enough.

> For the cross-platform case, I think the answer is to prefer running IWYU on a case-insensitive platform.

Hmm… You probably meant "on a case-sensitive platform"…? *(otherwise I don't get it)*

> But we might also want to consider letting users choose case sensitive or not for Windows, and just default to case-sensitive.

Even if you only consider case-insensitive platforms, you may find that people often use a mix of upper/down case letters. Just because it's easier to read this way. E.g. `QFileSystemModel` is simply more readable than `qfilesystemmodel`.

So, I think, the best thing that can be done to satisfy people on case-insensitive platforms, is an opt-in configuration to describe text transformation to apply to headers.

**Comment by kimgr** [comment]
> > For the cross-platform case, I think the answer is to prefer running IWYU on a case-insensitive platform.
>
> Hmm… You probably meant "on a case-sensitive platform"…? (otherwise I don't get it)

Oops. yes!

> So, I think, the best thing that can be done to satisfy people on case-insensitive platforms, is
> an opt-in configuration to describe text transformation to apply to includes.

Yes, I think that's what I suggested above.
